coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,181 @@
+		opt(hb)
+	}
+	if hb.tickCh == nil && hb.stopTick == nil {
+		ticker := time.NewTicker(DefaultFlushInterval)
+		hb.tickCh = ticker.C
+		hb.stopTick = ticker.Stop
+	}
+	return hb
+}
+
+type Option func(*Tracker)
+
+// WithLogger sets the logger to be used by Tracker.
+func WithLogger(log slog.Logger) Option {
+	return func(h *Tracker) {
+		h.log = log
+	}
+}
+
+// WithFlushInterval allows configuring the flush interval of Tracker.
+func WithFlushInterval(d time.Duration) Option {
+	return func(h *Tracker) {
+		ticker := time.NewTicker(d)
+		h.tickCh = ticker.C
+		h.stopTick = ticker.Stop
+	}
+}
+
+// WithFlushChannel allows passing a channel that receives
+// the number of marked workspaces every time Tracker flushes.
+// For testing only.
+func WithFlushChannel(c chan int) Option {
+	return func(h *Tracker) {
+		h.flushCh = c
+	}
+}
+
+// WithTickChannel allows passing a channel to replace a ticker.
+// For testing only.
+func WithTickChannel(c chan time.Time) Option {
+	return func(h *Tracker) {
+		h.tickCh = c
+		h.stopTick = func() {}
+	}
+}
+
+// Add marks the workspace with the given ID as having been used recently.
+// Tracker will periodically flush this to its configured Store.
+func (wut *Tracker) Add(workspaceID uuid.UUID) {
+	wut.mut.Lock()
+	wut.m[workspaceID] = struct{}{}
+	wut.mut.Unlock()
+}
+
+// flushLocked updates last_used_at of all current workspace IDs.
+// MUST HOLD LOCK BEFORE CALLING
+func (wut *Tracker) flushLocked(now time.Time) {
+	if wut.mut.TryLock() {
+		panic("developer error: must lock before calling flush()")
+	}
+	count := len(wut.m)
+	defer func() { // only used for testing
+		if wut.flushCh != nil {
+			wut.flushCh <- count
+		}
+	}()
+	if count == 0 {
+		wut.log.Debug(context.Background(), "nothing to flush")
+		return
+	}
+	// Copy our current set of IDs
+	ids := make([]uuid.UUID, 0)
+	for k := range wut.m {
+		ids = append(ids, k)
+	}
+	// Reset our internal map
+	wut.m = make(map[uuid.UUID]struct{})
+	// For ease of testing, sort the IDs lexically
+	sort.Slice(ids, func(i, j int) bool {
+		// For some unfathomable reason, byte arrays are not comparable?
+		return strings.Compare(ids[i].String(), ids[j].String()) < 0
+	})
+	// Set a short-ish timeout for this. We don't want to hang forever.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	// nolint: gocritic // system function
+	authCtx := dbauthz.AsSystemRestricted(ctx)
+	if err := wut.s.BatchUpdateWorkspaceLastUsedAt(authCtx, database.BatchUpdateWorkspaceLastUsedAtParams{
+		LastUsedAt: now,
+		IDs:        ids,
+	}); err != nil {
+		wut.log.Error(ctx, "failed updating workspaces last_used_at", slog.F("count", count), slog.Error(err))
+		return
+	}
+	wut.log.Info(ctx, "updated workspaces last_used_at", slog.F("count", count), slog.F("now", now))
+}
+
+func (wut *Tracker) Loop() {
+	defer func() {
+		wut.log.Debug(context.Background(), "workspace usage tracker loop exited")
+	}()
+	for {
```

Comment:
We could protect this from being called on a closed tracker. I think it could be a good protection from the future if someone just assumes they can "reopen" it.

```golang
	select { // Should we check the done channel to?
	case <-wut.stopCh:
		return
	default:
	}
```

Reply:
Gonna be even more strict and panic in this case.